shadowsocks/config.go

```diff
@@ -0,0 +1,149 @@
+// Copyright 2021 The Outline Authors
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//      http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package shadowsocks
+
+import (
+	"context"
+	"crypto/sha256"
+	"crypto/tls"
+	"crypto/x509"
+	"encoding/base64"
+	"encoding/json"
+	"errors"
+	"net"
+	"net/http"
+	"time"
+)
+
+// ProxyConfig represents a Shadowsocks proxy configuration.
+type ProxyConfig struct {
+	Host       string `json:"server"`
+	Port       int    `json:"server_port"`
+	Password   string `json:"password"`
+	Cipher     string `json:"method"`
+	Name       string `json:"remarks,omitempty"`
+	Plugin     string `json:"plugin,omitempty"`
+	PluginOpts string `json:"plugin_opts,omitempty"`
+}
+
+// FetchConfigRequest encapsulates a request to an online config server.
+type FetchConfigRequest struct {
+	// URL is the HTTPs endpoint of an online config server.
+	URL string
+	// Method is the HTTP method to use in the request.
+	Method string
+	// TrustedCertFingerprint is the base64-encoded sha256 hash of the online
```

Comment:
Currently, it looks like you're storing the fingeprint as base64, computing the new fingerprint in base64, and comparing them as strings.  This seems surprising to me.  Can we make this a `[]byte`, and do the comparison on the binary hashes?

Reply:
Done in eebc114. Good suggestion.